File: logservice/tests/builder.test.ts
import { describe, it, expect } from "bun:test";
import { buildOrderBy, buildWhere, parseSearchQuery } from "../src/query/builder";

const FIELDS = new Set(["sender", "dt", "host", "delay", "tls"]);

describe("buildWhere", () => {
    it("returns empty sql when no params", () => {
        const { sql, values } = buildWhere([], "AND", FIELDS);
        expect(sql).toBe("");
        expect(values).toEqual([]);
    });

    it("handles 'is' operator", () => {
        const { sql, values } = buildWhere(
            [{ field: "sender", operator: "is", value: "user@example.com" }],
            "AND", FIELDS
        );
        expect(sql).toBe("`sender` = ?");
        expect(values).toEqual(["user@example.com"]);
    });

    it("handles 'contains' operator", () => {
        const { sql, values } = buildWhere(
            [{ field: "sender", operator: "contains", value: "gmail" }],
            "AND", FIELDS
        );
        expect(sql).toBe("`sender` LIKE ?");
        expect(values).toEqual(["%gmail%"]);
    });

    it("handles 'begins' operator", () => {
        const { sql, values } = buildWhere(
            [{ field: "sender", operator: "begins", value: "user" }],
            "AND", FIELDS
        );
        expect(sql).toBe("`sender` LIKE ?");
        expect(values).toEqual(["user%"]);
    });

    it("handles 'ends' operator", () => {
        const { sql, values } = buildWhere(
            [{ field: "sender", operator: "ends", value: ".com" }],
            "AND", FIELDS
        );
        expect(sql).toBe("`sender` LIKE ?");
        expect(values).toEqual(["%.com"]);
    });

    it("handles 'between' operator", () => {
        const { sql, values } = buildWhere(
            [{ field: "dt", operator: "between", value: [1000, 2000] }],
            "AND", FIELDS
        );
        expect(sql).toBe("`dt` BETWEEN ? AND ?");
        expect(values).toEqual([1000, 2000]);
    });

    it("handles '>' and 'more' operators", () => {
        const r1 = buildWhere([{ field: "delay", operator: ">",    value: 5 }], "AND", FIELDS);
        const r2 = buildWhere([{ field: "delay", operator: "more", value: 5 }], "AND", FIELDS);
        expect(r1.sql).toBe("`delay` > ?");
        expect(r2.sql).toBe("`delay` > ?");
        expect(r1.values).toEqual([5]);
    });

    it("handles '<' and 'less' operators", () => {
        const r1 = buildWhere([{ field: "delay", operator: "<",    value: 5 }], "AND", FIELDS);
        const r2 = buildWhere([{ field: "delay", operator: "less", value: 5 }], "AND", FIELDS);
        expect(r1.sql).toBe("`delay` < ?");
        expect(r2.sql).toBe("`delay` < ?");
    });

    it("joins multiple conditions with AND", () => {
        const { sql, values } = buildWhere(
            [
                { field: "sender", operator: "contains", value: "gmail" },
                { field: "tls",    operator: "is",       value: 1 },
            ],
            "AND", FIELDS
        );
        expect(sql).toBe("`sender` LIKE ? AND `tls` = ?");
        expect(values).toEqual(["%gmail%", 1]);
    });

    it("joins multiple conditions with OR", () => {
        const { sql } = buildWhere(
            [
                { field: "sender", operator: "is", value: "a@a.com" },
                { field: "host",   operator: "is", value: "smtp.example.com" },
            ],
            "OR", FIELDS
        );
        expect(sql).toBe("`sender` = ? OR `host` = ?");
    });

    it("silently drops fields not in allowlist", () => {
        const { sql, values } = buildWhere(
            [{ field: "password", operator: "is", value: "secret" }],
            "AND", FIELDS
        );
        expect(sql).toBe("");
        expect(values).toEqual([]);
    });

    it("silently drops params with empty value", () => {
        const { sql } = buildWhere(
            [{ field: "sender", operator: "is", value: "" }],
            "AND", FIELDS
        );
        expect(sql).toBe("");
    });

    // The join operator used to be interpolated straight from caller-supplied
    // JSON, so `searchLogic: "OR 1=1 OR"` was a filter-bypass SQL injection.
    const TWO_PARAMS = [
        { field: "sender", operator: "is", value: "a@a.com" },
        { field: "host",   operator: "is", value: "smtp.example.com" },
    ] as const;

    it("never interpolates a hostile logic value into the sql", () => {
        const { sql, values } = buildWhere(
            [...TWO_PARAMS],
            // biome-ignore lint/suspicious/noExplicitAny: testing bad input
            "OR 1=1 OR" as any,
            FIELDS
        );
        expect(sql).toBe("`sender` = ? AND `host` = ?");
        expect(sql).not.toContain("1=1");
        expect(values).toEqual(["a@a.com", "smtp.example.com"]);
    });

    it("normalises the logic operator case-insensitively", () => {
        for (const logic of ["or", "Or", "OR"]) {
            // biome-ignore lint/suspicious/noExplicitAny: testing bad input
            const { sql } = buildWhere([...TWO_PARAMS], logic as any, FIELDS);
            expect(sql).toBe("`sender` = ? OR `host` = ?");
        }
    });

    it("falls back to AND for junk, empty and missing logic", () => {
        // biome-ignore lint/suspicious/noExplicitAny: testing bad input
        for (const logic of ["junk", "", null, undefined] as any[]) {
            const { sql } = buildWhere([...TWO_PARAMS], logic, FIELDS);
            expect(sql).toBe("`sender` = ? AND `host` = ?");
        }
    });

    it("blocks subquery and time-based injection payloads", () => {
        const payloads = [
            "AND `id` IN (SELECT id FROM Users WHERE hash LIKE 'a%') AND",
            "AND (SELECT SLEEP(5)) AND",
            "UNION SELECT",
        ];
        for (const logic of payloads) {
            // biome-ignore lint/suspicious/noExplicitAny: testing bad input
            const { sql } = buildWhere([...TWO_PARAMS], logic as any, FIELDS);
            expect(sql).toBe("`sender` = ? AND `host` = ?");
        }
    });
});

describe("parseSearchQuery", () => {
    it("returns {} for missing or malformed input", () => {
        expect(parseSearchQuery(null)).toEqual({});
        expect(parseSearchQuery("")).toEqual({});
        expect(parseSearchQuery("{not json")).toEqual({});
    });

    it("returns {} for non-object json", () => {
        expect(parseSearchQuery("null")).toEqual({});
        expect(parseSearchQuery('"str"')).toEqual({});
        expect(parseSearchQuery("[1,2]")).toEqual({});
        expect(parseSearchQuery("42")).toEqual({});
    });

    it("strips a hostile searchLogic down to AND", () => {
        const q = parseSearchQuery('{"searchLogic":"OR 1=1 OR"}');
        expect(q.searchLogic).toBe("AND");
    });

    it("preserves a valid searchLogic and the rest of the query", () => {
        const q = parseSearchQuery(
            '{"searchLogic":"OR","limit":50,"offset":10,' +
                '"search":[{"field":"sender","operator":"is","value":"a@a.com"}]}'
        );
        expect(q.searchLogic).toBe("OR");
        expect(q.limit).toBe(50);
        expect(q.offset).toBe(10);
        expect(q.search).toEqual([
            { field: "sender", operator: "is", value: "a@a.com" },
        ]);
    });

    it("defaults searchLogic to AND when absent", () => {
        expect(parseSearchQuery('{"limit":10}').searchLogic).toBe("AND");
    });
});

describe("buildOrderBy", () => {
    it("returns the fallback when no sort given", () => {
        expect(buildOrderBy(undefined, FIELDS, "`id` DESC")).toBe("`id` DESC");
        expect(buildOrderBy([], FIELDS, "`id` DESC")).toBe("`id` DESC");
    });

    it("defaults direction to DESC", () => {
        expect(buildOrderBy([{ field: "dt" }], FIELDS, "`id` DESC")).toBe(
            "`dt` DESC"
        );
    });

    it("normalises asc/desc (case-insensitive) and ignores junk", () => {
        expect(
            // biome-ignore lint/suspicious/noExplicitAny: testing bad input
            buildOrderBy([{ field: "dt", direction: "ASC" as any }], FIELDS, "x")
        ).toBe("`dt` ASC");
        expect(
            // biome-ignore lint/suspicious/noExplicitAny: testing bad input
            buildOrderBy([{ field: "dt", direction: "junk" as any }], FIELDS, "x")
        ).toBe("`dt` DESC");
    });

    it("joins multiple sort columns in order", () => {
        const sql = buildOrderBy(
            [
                { field: "host", direction: "asc" },
                { field: "delay", direction: "desc" },
            ],
            FIELDS,
            "`id` DESC"
        );
        expect(sql).toBe("`host` ASC, `delay` DESC");
    });

    it("drops fields not in the allowlist, falling back if none remain", () => {
        expect(
            buildOrderBy([{ field: "password" }], FIELDS, "`id` DESC")
        ).toBe("`id` DESC");
        expect(
            buildOrderBy(
                [{ field: "password" }, { field: "dt", direction: "asc" }],
                FIELDS,
                "`id` DESC"
            )
        ).toBe("`dt` ASC");
    });

    it("qualifies columns with a table prefix for joins", () => {
        expect(
            buildOrderBy([{ field: "md5" }], new Set(["md5"]), "`h`.`id` DESC", "h")
        ).toBe("`h`.`md5` DESC");
    });
});

File: logservice-go/internal/query/builder_test.go
package query

import (
	"encoding/json"
	"reflect"
	"testing"
)

// params decodes a JSON search array the way a real request would, so the tests
// exercise the same any-typed values the handler sees rather than hand-built Go
// values that could never arrive.
func params(t *testing.T, raw string) []Param {
	t.Helper()
	var p []Param
	if err := json.Unmarshal([]byte(raw), &p); err != nil {
		t.Fatalf("bad test fixture %q: %v", raw, err)
	}
	return p
}

// Every operator's exact SQL and placeholder values. Ported from
// logservice/tests/builder.test.ts.
func TestBuildWhere_OperatorsRenderTheExpectedSQL(t *testing.T) {
	allowed := set("sender", "id", "dt")

	cases := []struct {
		name   string
		search string
		sql    string
		values []any
	}{
		{"is", `[{"field":"sender","operator":"is","value":"a@b.com"}]`,
			"`sender` = ?", []any{"a@b.com"}},
		{"equals sign", `[{"field":"sender","operator":"=","value":"a@b.com"}]`,
			"`sender` = ?", []any{"a@b.com"}},
		{"begins", `[{"field":"sender","operator":"begins","value":"foo"}]`,
			"`sender` LIKE ?", []any{"foo%"}},
		{"contains", `[{"field":"sender","operator":"contains","value":"foo"}]`,
			"`sender` LIKE ?", []any{"%foo%"}},
		{"ends", `[{"field":"sender","operator":"ends","value":"foo"}]`,
			"`sender` LIKE ?", []any{"%foo"}},
		{"between", `[{"field":"dt","operator":"between","value":[1,2]}]`,
			"`dt` BETWEEN ? AND ?", []any{int64(1), int64(2)}},
		{"greater than", `[{"field":"id","operator":">","value":5}]`,
			"`id` > ?", []any{int64(5)}},
		{"more", `[{"field":"id","operator":"more","value":5}]`,
			"`id` > ?", []any{int64(5)}},
		{"greater or equal", `[{"field":"id","operator":">=","value":5}]`,
			"`id` >= ?", []any{int64(5)}},
		{"less than", `[{"field":"id","operator":"<","value":5}]`,
			"`id` < ?", []any{int64(5)}},
		{"less", `[{"field":"id","operator":"less","value":5}]`,
			"`id` < ?", []any{int64(5)}},
		{"less or equal", `[{"field":"id","operator":"<=","value":5}]`,
			"`id` <= ?", []any{int64(5)}},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			got := BuildWhere(params(t, tc.search), "AND", allowed, "")
			if got.SQL != tc.sql {
				t.Errorf("SQL = %q, want %q", got.SQL, tc.sql)
			}
			if !reflect.DeepEqual(got.Values, tc.values) {
				t.Errorf("values = %#v, want %#v", got.Values, tc.values)
			}
		})
	}
}

func TestBuildWhere_UnknownOperatorEmitsNoClause(t *testing.T) {
	got := BuildWhere(params(t, `[{"field":"sender","operator":"regexp","value":"x"}]`),
		"AND", set("sender"), "")
	if got.SQL != "" {
		t.Errorf("SQL = %q, want empty — an unknown operator must emit nothing", got.SQL)
	}
}

// The silent skip. This is the documented behaviour AND the documented hazard:
// a field nobody allowlisted produces no filter at all, so the query returns
// every row rather than an error.
func TestBuildWhere_UnknownFieldIsSkippedSilently(t *testing.T) {
	got := BuildWhere(params(t, `[{"field":"password","operator":"is","value":"x"}]`),
		"AND", set("sender"), "")
	if got.SQL != "" {
		t.Errorf("SQL = %q, want empty", got.SQL)
	}
	if len(got.Values) != 0 {
		t.Errorf("values = %#v, want none", got.Values)
	}
}

// `if (!item.value && item.value !== 0)`: the JS falsy set, minus zero.
func TestBuildWhere_FalsyValuesAreSkippedExceptZero(t *testing.T) {
	allowed := set("tls", "sender")

	cases := []struct {
		name   string
		search string
		want   string
	}{
		{"empty string", `[{"field":"sender","operator":"is","value":""}]`, ""},
		{"null", `[{"field":"sender","operator":"is","value":null}]`, ""},
		{"false", `[{"field":"tls","operator":"is","value":false}]`, ""},
		{"zero is kept", `[{"field":"tls","operator":"is","value":0}]`, "`tls` = ?"},
		{"true is kept", `[{"field":"tls","operator":"is","value":true}]`, "`tls` = ?"},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			got := BuildWhere(params(t, tc.search), "AND", allowed, "")
			if got.SQL != tc.want {
				t.Errorf("SQL = %q, want %q", got.SQL, tc.want)
			}
		})
	}
}

// A `between` whose value is not a two-element array drops the whole condition
// rather than becoming a half-open range.
func TestBuildWhere_BetweenNeedsATwoElementArray(t *testing.T) {
	for _, raw := range []string{
		`[{"field":"dt","operator":"between","value":5}]`,
		`[{"field":"dt","operator":"between","value":[1]}]`,
		`[{"field":"dt","operator":"between","value":[1,2,3]}]`,
		`[{"field":"dt","operator":"between","value":"1,2"}]`,
	} {
		got := BuildWhere(params(t, raw), "AND", set("dt"), "")
		if got.SQL != "" {
			t.Errorf("%s: SQL = %q, want empty", raw, got.SQL)
		}
	}
}

func TestBuildWhere_JoinsWithAndOrOr(t *testing.T) {
	allowed := set("a", "b")
	search := `[{"field":"a","operator":"is","value":1},{"field":"b","operator":"is","value":2}]`

	if got := BuildWhere(params(t, search), "AND", allowed, ""); got.SQL != "`a` = ? AND `b` = ?" {
		t.Errorf("AND: SQL = %q", got.SQL)
	}
	if got := BuildWhere(params(t, search), "OR", allowed, ""); got.SQL != "`a` = ? OR `b` = ?" {
		t.Errorf("OR: SQL = %q", got.SQL)
	}
}

// The injection payloads from logservice/tests/builder.test.ts:113-159. The
// joiner is re-normalised at SQL assembly, so nothing but a literal "OR" can
// produce an OR and everything else degrades to the more restrictive AND.
func TestBuildWhere_SearchLogicCannotInject(t *testing.T) {
	allowed := set("a", "b")
	search := `[{"field":"a","operator":"is","value":1},{"field":"b","operator":"is","value":2}]`

	for _, logic := range []string{
		"OR 1=1 OR",
		"AND (SELECT 1 FROM Users) AND",
		"AND SLEEP(5) AND",
		"UNION SELECT * FROM Users",
		"; DROP TABLE Users; --",
		"",
		"and",
		"nonsense",
	} {
		got := BuildWhere(params(t, search), logic, allowed, "")
		if got.SQL != "`a` = ? AND `b` = ?" {
			t.Errorf("logic %q produced %q, want the AND form", logic, got.SQL)
		}
	}
}

func TestBuildWhere_SearchLogicIsCaseInsensitiveForOr(t *testing.T) {
	allowed := set("a", "b")
	search := `[{"field":"a","operator":"is","value":1},{"field":"b","operator":"is","value":2}]`
	for _, logic := range []string{"or", "Or", "OR"} {
		if got := BuildWhere(params(t, search), logic, allowed, ""); got.SQL != "`a` = ? OR `b` = ?" {
			t.Errorf("logic %q produced %q, want the OR form", logic, got.SQL)
		}
	}
}

func TestBuildWhere_TablePrefixQualifiesColumns(t *testing.T) {
	got := BuildWhere(params(t, `[{"field":"md5","operator":"is","value":"abc"}]`),
		"AND", set("md5"), "h")
	if got.SQL != "`h`.`md5` = ?" {
		t.Errorf("SQL = %q, want `h`.`md5` = ?", got.SQL)
	}
}

// A number in a LIKE pattern must render the way JavaScript interpolated it —
// "42", never "42.000000".
func TestBuildWhere_NumericLikeValuesRenderWithoutTrailingZeros(t *testing.T) {
	got := BuildWhere(params(t, `[{"field":"a","operator":"contains","value":42}]`),
		"AND", set("a"), "")
	if len(got.Values) != 1 || got.Values[0] != "%42%" {
		t.Errorf("values = %#v, want [%%42%%]", got.Values)
	}
}

func TestBuildOrderBy(t *testing.T) {
	allowed := set("id", "dt")

	cases := []struct {
		name string
		sort string
		want string
	}{
		{"empty falls back", `[]`, "`id` DESC"},
		{"defaults to DESC", `[{"field":"dt"}]`, "`dt` DESC"},
		{"asc is honoured", `[{"field":"dt","direction":"asc"}]`, "`dt` ASC"},
		{"uppercase ASC", `[{"field":"dt","direction":"ASC"}]`, "`dt` ASC"},
		{"junk direction is DESC", `[{"field":"dt","direction":"; DROP TABLE"}]`, "`dt` DESC"},
		{"unknown field falls back", `[{"field":"password"}]`, "`id` DESC"},
		{"multi column", `[{"field":"dt","direction":"asc"},{"field":"id"}]`, "`dt` ASC, `id` DESC"},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			var sorts []Sort
			if err := json.Unmarshal([]byte(tc.sort), &sorts); err != nil {
				t.Fatalf("bad fixture: %v", err)
			}
			if got := BuildOrderBy(sorts, allowed, "`id` DESC", ""); got != tc.want {
				t.Errorf("BuildOrderBy = %q, want %q", got, tc.want)
			}
		})
	}
}

func TestBuildOrderBy_PrefixQualifies(t *testing.T) {
	sorts := []Sort{{Field: "md5", Direction: "asc"}}
	if got := BuildOrderBy(sorts, set("md5"), "`h`.`id` DESC", "h"); got != "`h`.`md5` ASC" {
		t.Errorf("BuildOrderBy = %q", got)
	}
}

// Anything that is not a JSON object yields the defaults — never an error, so a
// malformed q shows the console a first page rather than a broken grid.
func TestParse_NonObjectsDegradeToDefaults(t *testing.T) {
	for _, raw := range []string{"", "null", "[]", `"x"`, "5", "{not json", `[1,2,3]`} {
		q := Parse(raw)
		if len(q.Search) != 0 || q.SearchLogic != "" || len(q.Sort) != 0 {
			t.Errorf("Parse(%q) = %#v, want the zero Query", raw, q)
		}
		if limit, offset, _ := q.LimitOffset(); limit != DefaultLimit || offset != 0 {
			t.Errorf("Parse(%q) limit/offset = %d/%d, want %d/0", raw, limit, offset, DefaultLimit)
		}
	}
}

func TestParse_ReadsAWellFormedQuery(t *testing.T) {
	q := Parse(`{"limit":50,"offset":10,"searchLogic":"OR","search":[{"field":"a","operator":"is","value":"b"}],"sort":[{"field":"id","direction":"asc"}]}`)
	limit, offset, clamped := q.LimitOffset()
	if limit != 50 || offset != 10 || clamped {
		t.Errorf("limit/offset/clamped = %d/%d/%v, want 50/10/false", limit, offset, clamped)
	}
	if q.Logic() != "OR" {
		t.Errorf("Logic = %q, want OR", q.Logic())
	}
	if len(q.Search) != 1 || q.Search[0].Field != "a" {
		t.Errorf("Search = %#v", q.Search)
	}
}

// The one behaviour this port deliberately changes: the Bun service put q.limit
// straight into LIMIT with no ceiling.
func TestLimitOffset_IsClampedRatherThanRejected(t *testing.T) {
	cases := []struct {
		name        string
		raw         string
		wantLimit   int
		wantOffset  int
		wantClamped bool
	}{
		{"absent", `{}`, DefaultLimit, 0, false},
		{"in range", `{"limit":250}`, 250, 0, false},
		{"at the cap", `{"limit":1000}`, MaxLimit, 0, false},
		{"over the cap", `{"limit":100000000}`, MaxLimit, 0, true},
		{"negative", `{"limit":-5}`, DefaultLimit, 0, true},
		{"zero is honoured", `{"limit":0}`, 0, 0, false},
		{"non-numeric", `{"limit":"lots"}`, DefaultLimit, 0, false},
		{"negative offset", `{"offset":-5}`, DefaultLimit, 0, false},
		{"offset kept", `{"offset":40}`, DefaultLimit, 40, false},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			limit, offset, clamped := Parse(tc.raw).LimitOffset()
			if limit != tc.wantLimit || offset != tc.wantOffset || clamped != tc.wantClamped {
				t.Errorf("limit/offset/clamped = %d/%d/%v, want %d/%d/%v",
					limit, offset, clamped, tc.wantLimit, tc.wantOffset, tc.wantClamped)
			}
		})
	}
}

// Regression guard ported from logservice/tests/searchfields.test.ts.
func TestAllowlists_CarryTheGatewayAndRouteRuleColumns(t *testing.T) {
	for name, fields := range map[string]map[string]struct{}{
		"Delivery":    DeliveryFields,
		"Connection":  ConnectionFields,
		"Transaction": TransactionFields,
	} {
		if _, ok := fields["gateway"]; !ok {
			t.Errorf("%s allowlist is missing `gateway` — the filter would silently return every row", name)
		}
	}

	if _, ok := DeliveryFields["route_rule"]; !ok {
		t.Error("Delivery allowlist is missing `route_rule`")
	}
	// route_rule exists only on Delivery: routing is per recipient, and a
	// Transaction row is one message.
	if _, ok := TransactionFields["route_rule"]; ok {
		t.Error("Transaction allowlist has `route_rule`, which is not a column on that table")
	}
	if _, ok := ConnectionFields["route_rule"]; ok {
		t.Error("Connection allowlist has `route_rule`, which is not a column on that table")
	}
}
